### PR TITLE
[JENKINS-35451] - prefer ‘agent.jar’ over ‘slave.jar’

### PR DESCRIPTION
**Pull request by ndeloof**
See [JENKINS-35451](https://issues.jenkins-ci.org/browse/JENKINS-35451).

Proposed fix do still fully support use of `slave.jar` for existing automation but do also support `agent.jar` and use this name in help tips as well as web UI sample commands to run an agent.

### Proposed changelog entries

* Document Jenkins agent runtime as `agent.jar` and deprecate (but still support) use of legacy `slave.jar`, introduce the `AGENTJAR_URL` environment variable

### Desired reviewers

@oleg-nenashev @bitwiseman

### CODE REVIEW

File: core/src/main/java/hudson/slaves/CommandLauncher.java [hunk 1]
```diff
@@ -114,7 +114,7 @@ public void launch(SlaveComputer computer, final TaskListener listener) {
                 if (rootUrl!=null) {
                     pb.environment().put("HUDSON_URL", rootUrl);    // for backward compatibility
                     pb.environment().put("JENKINS_URL", rootUrl);
-                    pb.environment().put("SLAVEJAR_URL", rootUrl+"/jnlpJars/slave.jar");
+                    pb.environment().put("SLAVEJAR_URL", rootUrl+"/jnlpJars/agent.jar");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/CommandLauncher.java]: Maybe also `AGENTJAR_URL` though I do not see a need in this var at all

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 2]
```diff
@@ -521,7 +521,7 @@ public void onClosed(Channel c, IOException cause) {
             channel.addListener(listener);
 
         String slaveVersion = channel.call(new SlaveVersion());
-        log.println("Slave.jar version: " + slaveVersion);
+        log.println("agent.jar version: " + slaveVersion);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: Maybe "Remoting version: " ?
We cannot guarantee the name of the file in this place (actually there may be no file at all)

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 3]
```diff
@@ -362,7 +362,7 @@ public void setChannel(InputStream in, OutputStream out, TaskListener taskListen
      * Creates a {@link Channel} from the given stream and sets that to this agent.
      *
      * @param in
-     *      Stream connected to the remote "slave.jar". It's the caller's responsibility to do
+     *      Stream connected to the remote "agent.jar". It's the caller's responsibility to do
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: "To the remote agent"

File: core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html [hunk 4]
```diff
@@ -4,29 +4,29 @@
 
   <p>
   この項目にコマンドが指定されたとき、マスターでこのコマンドが実行されます。そして、
-  Jenkinsは、その実行コマンドが正しいスレーブマシーン上で<tt>slave.jar</tt>を起動することを想定しています。
+  Jenkinsは、その実行コマンドが正しいスレーブマシーン上で<tt>agent.jar</tt>を起動することを想定しています。
 
   <p>
-  <tt>slave.jar</tt>のコピーが、<tt>jenkins.war</tt>の中の<a href="${rootURL}/jnlpJars/slave.jar"><tt>WEB-INF/slave.jar</tt></a>にあります。
+  <tt>agent.jar</tt>のコピーが、<tt>jenkins.war</tt>の中の<a href="${rootURL}/jnlpJars/agent.jar"><tt>WEB-INF/agent.jar</tt></a>にあります。
```
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html]: There is no `WEB-INF/agent.jar` or slave.jar anymore. A while ago @jglick has removed these duplications, so there is only WEB-INF/lib/remoting-*.jar

Ideally we need to remove the reference from all translations

**Review by oleg-nenashev — APPROVED** [review]
Thanks for doing it @ndeloof! I added comments, but they are mostly related to the the previous code. So generally it's 🐝 , but it would be great if you could cleanup some things while you are around at the same codelines.

I also wonder what has changed in `test/src/test/resources/hudson/model/update-center.json`

File: core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html (hunk 4 above)
Comment (ndeloof) [reply core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html]: indeed. I'll need assistance to fix core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ru.html and core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html

File: core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ru.html [hunk 5]
```diff
@@ -4,34 +4,34 @@
 
   <p>
   Когда в этом поле указана команда, она будет выполнена на мастере и Jenkins предполагает
-  что эта команда запустит <tt>slave.jar</tt> на соответствующем подчиненном узле.
+  что эта команда запустит <tt>agent.jar</tt> на соответствующем подчиненном узле.
 
   <p>
-  Копия файла <tt>slave.jar</tt> вы можете найти в <a href="${rootURL}/jnlpJars/slave.jar"><tt>WEB-INF/slave.jar</tt></a> внутри 
+  Копия файла <tt>agent.jar</tt> вы можете найти в <a href="${rootURL}/jnlpJars/agent.jar"><tt>WEB-INF/agent.jar</tt></a> внутри
```
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ru.html]: `Копию файла <tt>agent.jar</tt> вы можете загрузить по <a href="${rootURL}/jnlpJars/agent.jar">этой ссылке</a>.`

there was also a typo

File: core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html (hunk 4 above)
Comment (oleg-nenashev) [reply core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html]: suggested Russian translation.
We need @kohsuke or @ikedam for the Japanese one

File: core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ru.html (hunk 5 above)
Comment (ndeloof) [reply core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ru.html]: thanks a lot ! I assume you don't speak Japanese :P

**Comment by jtnord** [comment]
just wondering how this will interact with things like the "install windows service" after an windows JNLP agent has been created (I think the service will reference the `slave.jar` yet that will no longer be present for new installs...  So possibly a follow up PR for the windows agent thingy that needs to link here.

**Comment by ndeloof** [comment]
IIUC slave installer do rely on https://github.com/jenkinsci/slave-installer-module/blob/master/src/main/java/org/jenkinsci/modules/slave_installer/impl/JnlpLaunchConfiguration.java#L36 to lookup the `remoting.jar` which is running the installer, so will just pick-up the expected one.

**Comment by oleg-nenashev** [comment]
Yep. Windows Slave Installer should work as well, we have already fixed that :(

**Comment by daniel-beck** [comment]
> Windows Slaves plugin

You mean Windows Agents plugin? 😉

**Comment by oleg-nenashev** [comment]
@daniel-beck There is https://github.com/jenkinsci/windows-slaves-plugin/pull/7 for it. We have not yet sent our DevOps Queen (aka Breaker of Chains) to that plugin.

**Review by bitwiseman — APPROVED** [review]
Yay!

**Comment by oleg-nenashev** [comment]
Still needs Japanese translation

**Comment by oleg-nenashev** [comment]
OTOH this PR soes not make it worse. I will create a follow-up defect and merge this pr

**Comment by oleg-nenashev** [comment]
Created https://issues.jenkins-ci.org/browse/JENKINS-46772

**Comment by oleg-nenashev** [comment]
Merging without squash, because commits are well separated. Fix for Russian localization may be backported, though it is not very important

File: core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html (hunk 4 above)
Comment (ikedam) [reply core/src/main/resources/hudson/slaves/CommandLauncher/help-command_ja.html]: Created jenkinsci/command-launcher-plugin#2

**Pull request merged**